src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.NumericValueSet.cs

```diff
@@ -97,260 +103,231 @@ static bool anyInterval(Interval interval, BinaryOperatorKind relation, T value,
-                        default:
-                            throw new ArgumentException(nameof(interval));
+                        builder.Add((tc.Next(_intervals[i].last), tc.Prev(_intervals[i + 1].first)));
+                    }
+                    // add a suffix if apropos
+                    if (tc.Related(LessThan, _intervals[_intervals.Length - 1].last, tc.MaxValue))
+                    {
+                        builder.Add((tc.Next(_intervals[_intervals.Length - 1].last), tc.MaxValue));
                     }
                 }
+
+                return new NumericValueSet<T, TTC>(builder.ToImmutableAndFree());
             }
 
             IValueSet IValueSet.Complement() => this.Complement();
 
             public IValueSet<T> Intersect(IValueSet<T> o)
             {
-                var other = ((NumericValueSet<T, TTC>)o);
-                var newInterval = intersectInterval(_rootInterval, other._rootInterval);
-                if (newInterval == _rootInterval)
-                    return this;
-                if (newInterval == other._rootInterval)
-                    return other;
-                return new NumericValueSet<T, TTC>(newInterval);
-
-                static Interval intersectInterval(Interval left, Interval right)
+                var other = (NumericValueSet<T, TTC>)o;
+                TTC tc = default;
+                var builder = ArrayBuilder<(T first, T last)>.GetInstance();
+                var left = this._intervals;
+                var right = other._intervals;
+                int l = 0;
+                int r = 0;
+                while (l < left.Length && r < right.Length)
                 {
-                    switch (left, right)
+                    var leftInterval = left[l];
+                    var rightInterval = right[r];
+                    if (tc.Related(LessThan, leftInterval.last, rightInterval.first))
                     {
-                        case (Interval.Excluded _, _):
-                            return left;
-                        case (_, Interval.Excluded _):
-                            return right;
-                        case (Interval.Included _, _):
-                            return right;
-                        case (_, Interval.Included _):
-                            return left;
-                        case (Interval.Mixed m1, Interval.Mixed m2):
-                            return intersectMixed(m1, m2);
-                        default:
-                            throw new ArgumentException("(left, right)");
+                        l++;
+                    }
+                    else if (tc.Related(LessThan, rightInterval.last, leftInterval.first))
+                    {
+                        r++;
+                    }
+                    else
+                    {
+                        Add(builder, Max(leftInterval.first, rightInterval.first), Min(leftInterval.last, rightInterval.last));
```

Comment:
Is it possible the interval is empty? If so, should that interval be dropped?

Reply:
No, it isn't possible, as the two if statements above ensure that the two intervals overlap.

---
In reply to: [401105048](https://github.com/dotnet/roslyn/pull/42886#discussion_r401105048) [](ancestors = 401105048)